### PR TITLE
Raster reprojection tutorial

### PR DESCRIPTION
**Pull request by petrsloup**
We were finally able to find enough time to finish the raster reprojection tutorial (related to #4122).

This PR adds raster reprojection tutorial explaining basic usage, inner workings and some advanced features of the raster reprojection. (no functional change)

Until now, there were no images (or other resources) for the tutorials.
I created `raster-reprojection-resources` folder for the required 3 images for this tutorial. Anybody has some better idea where to put these?
I'm not very happy with the long folder name, but none of the alternatives I could think of (`resources/raster-reprojection/___.jpg` or `resources/raster-reprojection-___.jpg`) was better.

### CODE REVIEW

File: doc/tutorials/raster-reprojection.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
+---
+title: Raster Reprojection
+layout: doc.hbs
+---
+
+# Raster Reprojection
+
+OpenLayers 3 has an ability to display raster data from WMS, WMTS, static images and many other sources in a different coordinate system than delivered from the server.
+Transformation of the map projections of the image happens directly in a web browser.
+The view in any Proj4js supported coordinate reference system is possible and previously incompatible layers can be now combined and overlaid.
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: can be now -> can now be

File: doc/tutorials/raster-reprojection.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+# Usage
+The usage in API is very simple. Just specify proper projection (using [EPSG](http://epsg.io) code) on `ol.View`:
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: usage in API -> API usage

File: doc/tutorials/raster-reprojection.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  }),
+  layers: [
+    new ol.layer.Tile({
+      source: new ol.source.TileWMS({
+        projection: 'EPSG:4326', //HERE IS THE DATA SOURCE PROJECTION
+        url: 'http://demo.boundlessgeo.com/geoserver/wms',
+        params: {
+          'LAYERS': 'ne:NE1_HR_LC_SR_W_DR'
+        }
+    })
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: indentation misses 2 spaces here?

File: doc/tutorials/raster-reprojection.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  })]
+});
+```
+If a source (based on `ol.source.TileImage` or `ol.source.Image`) has projection different from the current `ol.View`’s projection then the reprojection happens automatically under the hood.
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: has projection -> has a projection

File: doc/tutorials/raster-reprojection.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+### Examples
+- [Raster reprojection demo](http://openlayers.org/en/master/examples/reprojection.html)
+- [OpenStreetMap to WGS84 reprojection](http://openlayers.org/en/master/examples/reprojection-wgs84.html)
+- [Reprojection with EPSG.io database search](http://openlayers.org/en/master/examples/reprojection-by-code.html)
+- [Image reprojection](http://openlayers.org/en/master/examples/reprojection-image.html)
+
+### Custom projection
+The easiest way to use a custom projection is to add [Proj4js](http://proj4js.org/) library to your project and then define the projection using proj4 definition string.
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: to add -> to add the

Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: proj4 definition string -> a proj4 definition string

File: doc/tutorials/raster-reprojection.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+In certain cases, this can be used to optimize performance (by tweaking tile sizes) or visual quality (by specifying resolutions).
+
+# How it works
+
+The reprojection process is based on triangles -- the target raster is divided into a limited number of triangles with vertices transformed using `ol.proj` capabilities ([proj4js](http://proj4js.org/) is usually utilized to define custom transformations).
+The reprojection of pixels inside the triangle is approximated with an affine transformation (with rendering hardware-accelerated by the canvas 2d context):
+
+<img src="raster-reprojection-resources/how-it-works.jpg" alt="How it works" width="600" />
+
+This way we can support wide range of projections from proj4js (or even custom transformation functions) on almost any hardware (with canvas 2d support) with relatively small number of actual transformation calculations.
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: wide range -> a wide range

Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: with relatively small number -> with a relatively small number

File: doc/tutorials/raster-reprojection.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+The precision of the reprojection is then limited by the number of triangles.
+
+The reprojection process preserves transparency on the raster data supplied from the source (png or gif) and the gaps and no-data pixels generated by reprojection are automatically transparent.
+
+###Dynamic triangulation
+
+The above image above shows a noticeable error (especially on the edges) when the original image (left; EPSG:27700) is transformed with only a limited number of triangles (right; EPSG:3857).
+The error can be minimized by increasing the number of triangles used.
+
+Since some transformations require more detail triangulation network, the dynamic triangulation process automatically measures reprojection error and iteratively subdivides to meet a specific error threshold:
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: more detail -> a more detailed

File: doc/tutorials/raster-reprojection.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+<img src="raster-reprojection-resources/iterative-triangulation.png" alt="Iterative triangulation" width="600" />
+
+For debugging, rendering of the reprojection edges can be enabled by `ol.source.TileImage#setRenderReporojectionEdges(true)`.
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: setRenderReporojectionEdges -> setRenderReprojectionEdges

File: doc/tutorials/raster-reprojection.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+# Advanced
+
+### Disabling reprojection
+In case you are creating a custom build of OpenLayers and do not need the reprojection code, you can reduce the build size by setting `ol.ENABLE_RASTER_REPROJECTION` to `false`, which completely disables the reprojection support.
+See [Custom builds](custom-builds.html#defines) tutorial to see how to do this.
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: to see how -> on how

File: doc/tutorials/raster-reprojection.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+### Triangulation precision threshold
+The default [triangulation error threshold](#dynamic-triangulation) in pixels is given by `ol.DEFAULT_RASTER_REPROJECTION_ERROR_THRESHOLD` (0.5 pixel).
+In case different threshold needs to be defined for different sources, `reprojectionErrorThreshold` option can be passed when constructing tile image source.
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: different threshold -> a different threshold

Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: `reprojectionErrorThreshold` option -> the `reprojectionErrorThreshold` option

Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: constructing tile image source -> constructing the tile image source

File: doc/tutorials/raster-reprojection.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+###Limiting visibility of reprojected map by extent
+
+The reprojection algorithm uses inverse transformation (from *view projection* to *data projection*).
+For certain coordinate systems this can result in a "double occurrence" of the source data on a map.
+For example, when reprojecting a map of Switzerland from EPSG:21781 to EPSG:3857, it is displayed twice: once at the proper place in Europe, but also in the Pacific Ocean near the New Zealand, on the opposite side of the globe.
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: the New Zealand -> New Zealand

File: doc/tutorials/raster-reprojection.md [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+<img src="raster-reprojection-resources/double-occurrence.jpg" alt="Double occurrence of a reprojected map" width="600" />
+
+Although this is mathematically correct behavior of the inverse transformation, visibility of the map on multiple places is not expected by users.
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: visibility of the map -> visibility of the layer ?

File: doc/tutorials/raster-reprojection.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+A possible general solution would be to calculate the forward transformation for every vertex as well - but this would significantly decrease performance (especially for computationally expensive transformations).
+
+Therefore a recommended workaround is to define a proper visibility extent on the `ol.layer.Tile` in the view projection.
+Setting of such limit is demonstrated in the [reprojection demo example](http://openlayers.org/en/master/examples/reprojection.html).
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: Setting of such limit -> Setting such a limit

File: doc/tutorials/raster-reprojection.md [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+### Resolution calculation
+When determining source tiles to load, the ideal source resolution needs to be calculated.
+The `ol.reproj.calculateSourceResolution(sourceProj, targetProj, targetCenter, targetResolution)` function calculates the ideal value in order to achieve pixel mapping as close as possible to 1:1 during reprojection, which is then used to select proper zoom level from the source.
+
+It is, however, generally not practical to use the same source zoom level for the whole target zoom level -- different projections can have significantly different resolutions in different parts of the world (e.g. polar regions in EPSG:3857 vs EPSG:4326) and enforcing single resolution for whole zoom level would result in some tiles being scaled up/down, possibly requiring huge number of source tiles to be loaded.
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: enforcing single resolution -> enforcing a single resolution

Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: for whole zoom level -> for the whole zoom level

Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: requiring huge number -> requiring a huge number

File: doc/tutorials/raster-reprojection.md [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+Therefore, the resolution mapping is calculated for each reprojected tile separately (in the middle of the tile extent).
```
Comment (bartvde) [inline doc/tutorials/raster-reprojection.md]: calculated for each reprojected tile separately -> calculated separately for each reprojected tile

**Comment by bartvde** [comment]
thanks @petrsloup this looks great, I put a few small language remarks in the diff. Feel free to merge once these are addressed.

**Comment by petrsloup** [comment]
Thanks @bartvde for the remarks.

**Comment by klokan** [comment]
@bartvde when will this appear under /master/ on the website?

**Comment by bartvde** [comment]
I guess when somebody updates the website per https://github.com/openlayers/openlayers.github.io ;-)

Can you do it, or you want me to do it?

**Comment by bartvde** [comment]
I'm updating it now

**Comment by bartvde** [comment]
Actually there seems to be an error:

```
Assembling .grunt/openlayers-website/dist/en/master/doc/tutorials/raster-reprojection-resources/double-occurrence.html ERROR
Warning: Lexical error on line 1. Unrecognized text.
����JFIF���
----^ Use --force to continue.
```

**Comment by petrsloup** [comment]
@bartvde Thanks, I was just about to do it.
The error is in the build utilities (it tries to assemble `.html`s from images https://github.com/openlayers/openlayers.github.io/blob/build/gruntfile.js#L154). I'll try to fix it or create a ticket in that repo if I'm not able to.

**Comment by bartvde** [comment]
thanks @petrsloup

**Comment by petrsloup** [comment]
@bartvde PR with the fix: https://github.com/openlayers/openlayers.github.io/pull/55

**Pull request merged**